domain/secret/state/state.go

```diff
@@ -3154,9 +3155,65 @@ func (st State) DeleteSecret(ctx context.Context, uri *coresecrets.URI, revs []i
 	if err != nil {
 		return errors.Trace(err)
 	}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		return st.deleteSecretWithRevisions(ctx, tx, uri, revs)
+	})
+	return errors.Trace(domain.CoerceError(err))
+}
 
-	// First delete the specified revisions.
+// DeleteObsoleteUserSecretRevisions deletes the obsolete user secret revisions.
+func (st State) DeleteObsoleteUserSecretRevisions(ctx context.Context) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	q := `
+SELECT smo.secret_id AS &secretID.id,
+       sr.revision AS &secretExternalRevision.revision
+FROM   secret_model_owner smo
+       JOIN secret_metadata sm ON sm.secret_id = smo.secret_id
+       JOIN secret_revision sr ON sr.secret_id = smo.secret_id
+       LEFT JOIN secret_revision_obsolete sro ON sro.revision_uuid = sr.uuid
+WHERE  sm.auto_prune = true AND sro.obsolete = true`
+
+	stmt, err := st.Prepare(q, secretID{}, secretExternalRevision{}, revisions{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		var (
+			dbSecrets    secretIDs
+			dbsecretRevs secretExternalRevisions
+		)
+		err = tx.Query(ctx, stmt).GetAll(&dbSecrets, &dbsecretRevs)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			// Nothing to delete.
+			return nil
+		}
+		if err != nil {
+			return errors.Trace(err)
+		}
+		itemsToDelete, err := dbSecrets.toSecretMetadataForDrain(dbsecretRevs)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for _, toDelete := range itemsToDelete {
+			revs := make([]int, len(toDelete.Revisions))
+			for i, r := range toDelete.Revisions {
+				revs[i] = r.Revision
+			}
+			if err := st.deleteSecretWithRevisions(ctx, tx, toDelete.URI, revs); err != nil {
+				return errors.Trace(err)
+			}
+		}
+		return nil
+	})
+	return errors.Trace(domain.CoerceError(err))
+}
 
+func (st State) deleteSecretWithRevisions(ctx context.Context, tx *sqlair.TX, uri *coresecrets.URI, revs []int) error {
```

Comment:
The name of this method should be `deleteSecretRevisions`.
There should be a comment that if the secret ends up having no revisions left, the entire secret is removed.